### PR TITLE
Swift 4 compatibility hack for redeclaration of properties in generic type and extension

### PR DESCRIPTION
**Pull request by huonw**
The patch that nailed down our semantics here missed an additional case that
required a compatibility hack: a property on a generic type and a same-named one
in an (unconstrained) extension:

    struct Foo<T> {
        var x: Int { return 0 }
    }
    extension Foo {
        var x: Bool { return false }
    }

Fixes rdar://problem/40685642.

### CODE REVIEW

**Comment by huonw** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -1049,6 +1049,17 @@ static void checkRedeclaration(TypeChecker &tc, ValueDecl *current) {
           continue;
       }
 
+      // If both are VarDecls, and both have exactly the same type, then
+      // matching the Swift 4 behaviour (i.e. just emitting the future-compat
+      // warning) will result in SILGen crashes due to both properties mangling
+      // the same, so it's better to just follow the Swift 5 behaviour and emit
+      // the actual error.
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckDecl.cpp]: Can you add a test case for this?

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -1747,7 +1747,31 @@ bool swift::conflicting(ASTContext &ctx,
   }
 
   // Otherwise, the declarations conflict if the overload types are the same.
-  return sig1Type == sig2Type;
+  if (sig1Type == sig2Type) {
```
Comment (jrose-apple) [inline lib/AST/Decl.cpp]: Suggestion: early return.

**Review by jrose-apple — COMMENTED** [review]
Seems reasonable, if unfortunate.

File: lib/Sema/TypeCheckDecl.cpp (hunk 1 above)
Comment (huonw) [reply lib/Sema/TypeCheckDecl.cpp]: The last/unchanged property in the diff for the `overload.swift` is this case.

Comment (jrose-apple) [reply lib/Sema/TypeCheckDecl.cpp]: Oops, right, thanks.

**Comment by huonw** [comment]
@swift-ci please smoke test

**Pull request merged**